README.md
# Keyword Protocol 2000
The [KWP2000](https://en.wikipedia.org/wiki/Keyword_Protocol_2000/) is a communications protocol used for [on-board vehicle diagnostics systems (OBD)](https://en.wikipedia.org/wiki/On-board_diagnostics) by Suzuki (SDS), Kawasaki (KDS), Yamaha (YDS), Honda (HDS) and more. It is standardized by the [ISO 14230](https://www.iso.org/obp/ui/#iso:std:iso:14230:-1:ed-2:v1:en) and it is compatible with the [ISO 9141](https://www.iso.org/obp/ui/#iso:std:iso:9141:-2:ed-1:v1:en). Both uses a single line called the K-line through it we send some Parameter IDs, knowns as PIDs, to the [Electronic Control Unit (ECU)](https://en.wikipedia.org/wiki/Electronic_control_unit). This way we have total control on the motorbike: we are able to ask for sensors data (such as RPM, gear, speed, temperatures, etc.), see error codes and clear them, upload/download and more.

If you want to read more about this protocol and OBD-II you can see the [protocol explained](protocol.md).

### Hardware
Any microcontroller (MCU) could be used.

Many IC could be used as interface, for example:
- L9637
- MC33660
- MC33199
- operational amplifiers/level shifter/transistors (**not recommended**)

### List of Vehicles

| Motorbike     | Year      | Tested |
| ------------- |---------- |------- |
| GSX-R600      | 2011      | ✔️    |


# Usage
### Hardware
You have to find the K-line of your bike. In Suzuki bikes it's under the rider's seat. You just need to connect the Kline, Vcc and Gnd to the driver IC you bought

Example of the wiring for the L9636:

![Alt text](/extras/Images/l9637_wiring.png?raw=true "L9637 wiring")


##### Dealer Mode - Only for Suzuki:
If you wish also to be able to enter easily into this mode add an optocoupler with a ~330omh resistor between the MCU and the dealer pin

![Alt text](/extras/Images/dealer_mode.png?raw=true "dealer mode")


### Software
First of all go to [PIDs.h](/src/PIDs.h) and decomment (delete the `//` symbols) your motorbike, then upload any of the [examples](/examples/)


### Installation
Simply search for `KWP2000` in the Arduino/PlatformIO Library Manager or download this repository and add it to your library folder


### Development
I made a [ECU Emulator](/extras/ECU_Emulator) written in python for the development of new functions and tests.


### Documentation
Generally the functions return `true` if everything went fine, a `negative number` if there where any error, `false` if nothing changed

This documentation has been automatically generated with doxygen + moxygen, an automatic documentation generator, I will make the formatting nicer later.

See it here [documentation](documentation.md)


# Disclaimer
I do not assume and hereby disclaim any liability to any party for any damage to proprierty or person


# Influence:
- [ECU Hacking forum](https://ecuhacking.activeboard.com/): 
    - https://ecuhacking.activeboard.com/t56234221/kds-protocol
    - https://ecuhacking.activeboard.com/t22573776/sds-protocol/

- [Arduino forum](https://forum.arduino.cc/)
    - https://forum.arduino.cc/index.php?topic=236092.0

- Others


changelog.md
# Version - Change Log

## Future Development
- test on other bikes
- discover new PIDs

#### 1.1.0 - jan 13, 2019
- added `readTroubleCodes()` and `clearTroubleCodes()`
- small fixes

#### 1.0.0 - jan 10, 2019
- version promoted for pubblication on the arduino library manager
- code cleaner
- created documentation with doxygen
- removed the pointers complex code to a different branch
- removed the unofficial STM32 USBSerial
- code cleared

#### 0.5.1 - oct 28, 2018
- added handleRequest
- added changeTimingParameter
- better keepAlive
- solved bug in ECU emulator
- found [bug](https://github.com/espressif/arduino-esp32/issues/2004) in ESP32 core - waiting for someone who can fix it meanwhile I moved to STM32
- added clearError
- better checkResponse
- added resetTimingParameter

#### 0.5.0 - Oct 26, 2018
- completely rewritten the sendRequest, listenResponse and checkResponse
- more compliant with the ISO14230
- removed std::notrow since [arduino seems to not support it](https://github.com/arduino/ArduinoCore-avr/issues/47)
- added board/MCU specific flags at compile time
- added ability to read the keybytes and the timing parameters
- split the readme in multiple files for readability
- many improvements in the python ECU emulator

#### 0.4.0 - Oct 12, 2018
- made two examples sketches
- added ECU_Emulator written in python
- merged printStatus and printError
- added printSensorData
- added printLastResponse
- added new errors and better error handling
- added documentation
- review all the code
- added more resources

#### 0.3.0 - Sep 13, 2018
- solved the bug with FreeRTOS
- moved the BT serial in a fork
- added wonderful code to manage comunication errors

#### 0.2.0 - Aug 30, 2018
- problem with FreeRTOS 
- big cleaning of the code
- removed all the blocking code
- added BT serial

#### 0.1.0 - Mar 22, 2018
- created github repo


src/KWP2000.h
/*
KWP2000.h

Copyright (c) 2019 Aster94

Permission is hereby granted, free of charge, to any person obtaining a copy of this software 
and associated documentation files (the "Software"), to deal in the Software without restriction, 
including without limitation the rights to use, copy, modify, merge, publish, distribute, 
sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is 
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or 
substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING 
BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND 
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, 
DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

/*  TODO LIST 
check types
pids from manzo / ecuhacking / forum ?
make a two dimensional array to save errors?
remove resetError?
in keepalive should i try to send a request anyway instead of closing?
remove echo?
check with logic analyzer
check if te ecu is connected before some functions
check if all serial have F
check verbose/default levels

before commits:
update keywords.txt, changelog and library.properties
check pids.h and add the //
run doxygen + moxygen
*/

#ifndef KWP2000_h
#define KWP2000_h

/**
 * @brief Collection of possible debug levels
 */
enum debug_enum
{
    DEBUG_LEVEL_NONE,
    DEBUG_LEVEL_DEFAULT,
    DEBUG_LEVEL_VERBOSE
};

/**
 * @brief Used by `readTroubleCodes()`
 */
enum trouble_codes
{
    READ_TOTAL,
    READ_ONLY_ACTIVE,
    READ_ALL
};

class KWP2000
{
  public:
    // CONSTRUCTOR
    KWP2000(HardwareSerial *kline_serial, const uint8_t k_out_pin, const uint32_t kline_baudrate = 10400);

    // SETUP
    void enableDebug(HardwareSerial *debug_serial, const uint8_t debug_level = DEBUG_LEVEL_DEFAULT, const uint32_t debug_baudrate = 115200);
    void setDebugLevel(const uint8_t debug_level);
    void disableDebug();
    void enableDealerMode(const uint8_t dealer_pin);
    void dealerMode(const uint8_t dealer_mode);

    // COMMUNICATION - Basic
    int8_t initKline();
    int8_t stopKline();
    void requestSensorsData();
    void readTroubleCodes(const uint8_t which = READ_ONLY_ACTIVE);
    void clearTroubleCodes(const uint8_t code = 0x00);
    void keepAlive(uint16_t time = 0);

    // COMMUNICATION - Advanced
    int8_t handleRequest(const uint8_t to_send[], const uint8_t send_len, const uint8_t try_once = false);
    void accessTimingParameter(const uint8_t read_only = true);
    void resetTimingParameter();
    void changeTimingParameter(uint32_t new_atp[], const uint8_t new_atp_len);

    // PRINT and GET
    void printStatus(uint16_t time = 2000);
    void printSensorsData();
    void printLastResponse();
    int8_t getStatus();
    int8_t getError();
    void resetError();
    uint8_t getGPS();
    uint8_t getRPM();
    uint8_t getSPEED();
    uint8_t getTPS();
    uint8_t getIAP();
    uint8_t getIAT();
    uint8_t getECT();
    uint8_t getSTPS();

  private:
    // K-Line
    HardwareSerial *_kline;
    uint32_t _kline_baudrate;
    uint8_t _k_out_pin;
    uint8_t _dealer_pin;
    uint8_t _dealer_mode;
    uint8_t _init_sequence_started = false;
    uint8_t _stop_sequence_started = false;
    uint32_t _start_time = 0;
    uint32_t _elapsed_time = 0;
    uint8_t _response[260]; //todo use max data
    uint8_t _response_len = 0;
    uint8_t _response_data_start = 0;
    uint8_t _request[20];
    uint8_t _request_len = 0;
    uint8_t _ECU_status = false;
    uint32_t _ECU_error = 0;

    // k line config
    uint8_t _use_lenght_byte = true;
    uint8_t _use_target_source_address = true;
    uint8_t _timing_parameter = true; // normal
    uint16_t ISO_T_IDLE = 0;
    uint8_t ISO_T_P2_MIN = 25;
    uint32_t ISO_T_P2_MAX = 50;
    uint16_t ISO_T_P3_MIN = 55;
    uint32_t ISO_T_P3_MAX = 2000;
    uint32_t ISO_T_P3_mdf = 2000;
    uint16_t ISO_T_P4_MIN = 10; // average between min and max value
    uint16_t _keep_iso_alive = 1000;

    // debug
    HardwareSerial *_debug;
    uint8_t _debug_enabled = false;
    uint32_t _debug_baudrate;
    uint8_t _debug_level = DEBUG_LEVEL_DEFAULT;
    uint32_t _last_status_print = 0;
    uint32_t _last_data_print = 0;
    uint32_t _last_sensors_calculated = 0;
    uint32_t _last_correct_response = 0;
    uint32_t _connection_time = 0;

    // sensors
    uint8_t _GPS, _RPM, _SPEED, _TPS, _IAP, _ECT, _STPS, _IAT;
    uint8_t _GEAR1, _GEAR2, _GEAR3;

    // functions
    void sendRequest(const uint8_t to_send[], const uint8_t send_len, const uint8_t wait_to_send_all = true, const uint8_t use_delay = true);
    void listenResponse(const uint8_t use_delay = true);
    int8_t checkResponse(const uint8_t response_sent[]);
    void setError(const uint8_t error);
    void clearError(const uint8_t error);
    void configureKline();
    uint8_t calc_checksum(const uint8_t data[], const uint8_t data_len);
    void endResponse(const uint8_t received_checksum);
    void connectionExpired();
};

#endif // KWP2000_h


extras/ECU_Emulator/ECU_Emulator.py
# -*- coding: utf-8 -*-
"""
Created on Fri Aug  4 12:39:40 2017

@author: Vincenzo G.

module needed:
    pyserial
"""



import serial, time, sys
from random import randint

default_port = 'COM6' # 'COMx' for windows '/dev/ttyUSBx' for linux
this_address = 0xf1
ecu_address = 0x12
P2m = 40
P2M_ms = 1000
P3m = 50
P3M_ms = 1500
P4m = 10

if P2M_ms <= 6000:
    P2M_hex = int(P2M_ms / 25)
elif P2M_ms > 6000 and P2M_ms <= 89600:
    P2M_hex = int(P2M_ms / 256 / 25) | 0xf0
else:
    print ("P2M time too long")
    sys.exit()

if P3M_ms <= 6000:
    P3M_hex = int(P3M_ms / 25)
elif P3M_ms > 6000 and P3M_ms <= 89600:
    P3M_hex = int(P3M_ms / 256 / 25) | 0xf0
else:
    print ("P3M time too long")
    sys.exit()

start_com = [0x81, ecu_address, this_address, 0x81, 0x5]
start_com_ok = [0x80, this_address, ecu_address, 0x03, 0xC1, 0xEA, 0x8F]

stop_com = [[0x81, ecu_address, this_address, 0x82, 0x6],
            [0x80, ecu_address, this_address, 0x1, 0x82, 0x6]]
stop_com_ok = [0x80, this_address, ecu_address, 0x01, 0xC2]

atp_current = [[0x80, ecu_address, this_address, 0x2, 0x83, 0x02, 0xa],
               [0x82, ecu_address, this_address, 0x83, 0x02, 0xa]]
atp_current_ok = [0x80, this_address, ecu_address, 0x07 , 0xC3, 0x2, P2m, P2M_hex, P3m, P3M_hex, P4m]

atp_limits = [[0x80, ecu_address, this_address, 0x2, 0x83, 0x00, 0x8],
              [0x82, ecu_address, this_address, 0x83, 0x00, 0x8]]
atp_limits_ok = [0x80, this_address, ecu_address, 0x07 , 0xC3, 0x0, 1, 2, 3, 4, 1]

request_sens = [[0x80, ecu_address, this_address, 0x2, 0x21, 0x8, 0xae],
                [0x82, ecu_address, this_address, 0x21, 0x8, 0xae]]
fake_sens = [0x80, this_address, ecu_address, 0x34]

reject_request = [0x80, this_address, ecu_address, 0x01, 0x7F]

request = list()
ECU_connected = False
ATP_setted = False
ATP_asked = False
last_byte = 0
last_data_received = 0

ser = serial.Serial(
    port=default_port,
    baudrate=115200,
    parity=serial.PARITY_ODD,
    stopbits=serial.STOPBITS_ONE,
    bytesize=serial.EIGHTBITS,
    timeout = 2
    #timeout = None
)


# send an array of int as binary data
def send (to_send):
    to_send.append(calc_cs(to_send))
    #print("Sending:\t" + str([hex(x) for x in to_send]))
    for i in to_send:
        ser.write(i.to_bytes(1,byteorder='little'))
        time.sleep(P4m/1000.0)
    time.sleep(P2m/1000.0)


# compare two list
def compare(received,correct):
    #print (received)
    temp = list()
    if (type(correct[0]) != list):
        temp.append(correct)
    else:
        temp = correct
    for element in temp:
        coincident = 0
        l = len(element)
        for n in range(0,l):
            try:
                #print (element[n], int(received[n], 16))
                if (element[n] == int(received[n], 16)):
                    coincident += 1
            except IndexError:
                next
        if (coincident == l):
            return True
    return False


# calculate the checksum
def calc_cs (arr):
    return sum(arr) & 0xFF
    

# listen from the board 
def listen():
    global last_data_received
    byte = 0
    last_byte = -1
    request.clear()
    start_time = time.time()
    elapsed_time = 0

    while (ser.in_waiting > 0):
        elapsed_time = time.time() - start_time
        
        while (elapsed_time < (P3M_ms/1000.0)):
            in_bin = ser.read()
            in_hex = hex(int.from_bytes(in_bin, byteorder='little'))       
            #print ("Incoming byte " + str(byte) + ": " + in_hex)
            request.append(in_hex)
            
            # send back the byte we just received
            ser.write(in_bin)
            
            if byte == 0:
                if in_hex == '0x81':
                    # we will receive only 4 more bytes
                    last_byte = 4
                elif in_hex == '0x80':
                    # we don't know yet
                    last_byte = -1
            
            if byte == 3 and last_byte == -1:
                last_byte = int(in_hex, 0) + 4 #3 headers + 1 checksum
                
            if (byte == last_byte):
                # check the checksum
                cs = hex(calc_cs(int(x,0) for x in request[0:byte]))
                if in_hex == cs:
                    #print("request received: " + str(request))
                    last_data_received = time.time()
                    time.sleep(P2m/1000.0)
                    return True
                else:
                    print("Wrong checksum!")
                    print(request)
                    print("Expected: " + str(cs))
                    time.sleep(P2m/1000.0)
                    return False
            
            byte += 1
    return False


################### START ###################

print("Starting ECU Emulator")
try:
    if (ser.is_open):
        print("Port already opened")
    else:
        ser.open()
    print("Port Info:\n", str(ser))
except Exception as e:
    print ("Error open serial port: " + str(e))
    sys.exit()

while 1:
    # connection expired
  
    if ((time.time() - last_data_received) > (P3M_ms/1000.0)):
        if ECU_connected == True:
            print("Connection expired")
            print("From last data: " + str(P3M_ms/1000.0))
        ECU_connected = False
        ATP_setted = False
        ATP_asked = False
        request.clear()
       
    if (listen()): # we received something
        
        if not ECU_connected:
            if compare(request,start_com):
                print('\nConnection established\n')
                send(start_com_ok)
                ECU_connected = True   

            else:
                print('You should send me the init sequence! not this:')
                print(request)
                
        elif ECU_connected and not ATP_asked and not ATP_setted:
            if compare(request, atp_limits):
                print('ATP limits asked\n')
                send(atp_limits_ok)
                ATP_asked = True
                
            elif compare (request, stop_com):
                print("The tester closed the link")
                send(stop_com_ok)
                ECU_connected = False

            else:
                print('You should ask me the ATP limits! not this:')
                print(request)
        
        elif ECU_connected and ATP_asked and not ATP_setted:
            if compare(request,atp_current):
                print('Sending ATP response\n')
                send(atp_current_ok)
                ATP_setted = True
                
            elif compare (request, stop_com):
                print("The tester closed the link")
                send(stop_com_ok)
                ECU_connected = False
                
            else:
                print("Unkown request:")
                print(request)
        
        elif ECU_connected and ATP_setted:
            if compare(request,start_com):
                print("Already connected")
            
            elif compare(request,request_sens):
                print("Sending fake sensor data\n")
                for i in range(0,52):
                    fake_sens.append(randint(0,255))
                send(fake_sens)
                del fake_sens[4:]
            
            elif compare (request, stop_com):   
                print("The tester closed the link")
                send(stop_com_ok)
                ECU_connected = False
            
            else:
                print("Unkown request:")
                print(request)


src/PIDs.h
/*
PIDs.h
Parameter IDs

Copyright (c) 2019 Aster94

Permission is hereby granted, free of charge, to any person obtaining a copy of this software 
and associated documentation files (the "Software"), to deal in the Software without restriction, 
including without limitation the rights to use, copy, modify, merge, publish, distribute, 
sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is 
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or 
substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING 
BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND 
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, 
DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

////////////// MODIFY HERE ////////////////

// First choose only one of these motorbikes:
// To choose delete the "//" at the beginning of the line
//#define SUZUKI
//#define KAWASAKI
//#define YAMAYA
//#define HONDA

////////////// DON'T MODIFY UNDER THIS LINE ////////////////
////////////// UNLESS YOU KNOW WHAT YOU ARE DOING ////////////////

const uint8_t format_physical = 0x80;
const uint8_t format_functional = 0xC0; // not supported
const uint8_t format_CARB = 0x40;       // not supported
const uint8_t ECU_addr = 0x12;          // between hex 10 and 17
const uint8_t OUR_addr = 0xF1;          // between hex F0 and FD

const uint8_t start_com[] = {0x81};
const uint8_t stop_com[] = {0x82};

// access timing parameter and timing paramenter indicator
const uint8_t atp_read_limits[] = {0x83, 0x00};
const uint8_t atp_set_default[] = {0x83, 0x01};
const uint8_t atp_read_current[] = {0x83, 0x02};
const uint8_t atp_set_given[] = {0x83, 0x03};

const uint8_t tester_present_with_answer[] = {0x3E, 0x01};    // the ECU will ansker
const uint8_t tester_present_without_answer[] = {0x3E, 0x02}; // the ECU won't answer

const uint8_t trouble_codes_all[] = {0x13};
const uint8_t trouble_codes_only_active[] = {0x17};
const uint8_t trouble_codes_with_status[] = {0x18};

const uint8_t clear_trouble_codes[] = {0x14};

#define request_ok(x) x | 0x40
const uint8_t request_rejected = 0x7F;

#if defined(SUZUKI)
#if defined(BIKE_CHOOSED)
#error "two or more bikes choosed, please see src/PIDs.h"
#else
#define BIKE_CHOOSED
#endif // BIKE_CHOOSED

const uint8_t request_sens[] = {0x21, 0x08};

#define PID_SPEED 16
#define PID_RPM_H 17 // high byte
#define PID_RPM_L 18 // low byte
#define PID_TPS 19
#define PID_IAP 20
#define PID_ECT 21
#define PID_IAT 22
#define PID_AP 23
#define PID_BATT 24
#define PID_HO2 25
#define PID_GPS 26
//31 to 38 FUEL 1 2 3 H and L
//41 to 44 ignition IGN
#define PID_STPS 47
#define PID_PAIR 51
#define PID_CLUTCH 52
#define PID_GEAR_3 53

//from mark's (ciclegadget) PID_STP 46
#endif // SUZUKI

#if defined(KAWASAKI)
#if defined(BIKE_CHOOSED)
#error "two or more bikes choosed, please see src/PIDs.h"
#else
#define BIKE_CHOOSED
#endif // BIKE_CHOOSED

const uint8_t request_sens[][4] = {
    {0, 1, 2, 3},  //Speed
    {4, 5, 6, 7},  //RPM
    {8, 9, 10, 11} //GEAR
    //todo
};

#define PID_SPEED 16
#define PID_RPM_H 17 // high byte
#define PID_RPM_L 18 // low byte
#define PID_TPS 19
#define PID_IAP 20
#define PID_ECT 21
#define PID_IAT 22
#define PID_AP 23
#define PID_BATT 24
#define PID_HO2 25
#define PID_GPS 26
//31 to 38 FUEL 1 2 3 H and L
//41 to 44 ignition IGN
#define PID_STPS 47
#define PID_PAIR 51
#define PID_CLUTCH 52
#define PID_GEAR_3 53

#endif // KAWASAKI

#if defined(YAMAYA)
#if defined(BIKE_CHOOSED)
#error "two or more bikes choosed, please see src/PIDs.h"
#else
#define BIKE_CHOOSED
#endif // BIKE_CHOOSED
/*
As far as I could understand the YDS (Yamaha Diagnostic Protocol) protocol, it is close to the Honda protocol.
There are no sender-/receiver-adresses or header information. Just the checksum at the end.
You initialize the diagnostic mode by sending 0x80 and from then on you can just submit 0x02 and it gives you:
Rpm, Speed, Error, Gear & Checksum.
todo: need more info
*/
#endif // YAMAYA

#if defined(HONDA)
#if defined(BIKE_CHOOSED)
#error "two or more bikes choosed, please see src/PIDs.h"
#else
#define BIKE_CHOOSED
#endif // BIKE_CHOOSED
// see above
#endif // HONDA

#ifndef BIKE_CHOOSED
#error "no bike selected, please see src/PIDs.h"
#endif // BIKE_CHOOSED
